tcms/testruns/views.py

```diff
@@ -271,36 +272,71 @@ def _walk_executions(test_executions):
-    except ObjectDoesNotExist:
-        raise Http404
+    template_name = 'testruns/mutable.html'
+    form_class = BaseRunForm
+
+    def form_invalid(self, form):
+        test_run = self._get_test_run_or_fail(self.kwargs['run_id'])
+
+        context_data = {
+            'test_run': test_run,
+            'test_plan': test_run.plan,
+            'form': self._get_base_form(test_run),
+            'errors': form.errors
+        }
 
-    # If the form is submitted
-    if request.method == "POST":
-        form = BaseRunForm(request.POST)
+        return render(self.request, self.template_name, context_data)
+
+    def form_valid(self, form):
+        run_id = self.kwargs['run_id']
+        test_run = self._get_test_run_or_fail(run_id)
+        test_run.summary = form.cleaned_data['summary']
+        # Permission hack
+        if test_run.manager == self.request.user or test_run.plan.author == self.request.user:
+            test_run.manager = form.cleaned_data['manager']
+        test_run.default_tester = form.cleaned_data['default_tester']
+        test_run.build = form.cleaned_data['build']
+        test_run.product_version = test_run.plan.product_version
+        test_run.notes = form.cleaned_data['notes']
+        test_run.save()
+
+        return HttpResponseRedirect(reverse('testruns-get', args=[run_id, ]))
+
+    def get_form(self, form_class=BaseRunForm):
+        test_run = self._get_test_run_or_fail(self.kwargs['run_id'])
+        form = form_class(self.request.POST)
         form.populate(product_id=test_run.plan.product_id)
 
-        # FIXME: Error handler
-        if form.is_valid():
-            test_run.summary = form.cleaned_data['summary']
-            # Permission hack
-            if test_run.manager == request.user or test_run.plan.author == request.user:
-                test_run.manager = form.cleaned_data['manager']
-            test_run.default_tester = form.cleaned_data['default_tester']
-            test_run.build = form.cleaned_data['build']
-            test_run.product_version = test_run.plan.product_version
-            test_run.notes = form.cleaned_data['notes']
-            test_run.save()
-
-            return HttpResponseRedirect(reverse('testruns-get', args=[run_id, ]))
-    else:
+        return form
+
+    def get_context_data(self, **kwargs):
+        test_run = self._get_test_run_or_fail(self.kwargs['run_id'])
+        form = self._get_base_form(test_run)
+        form.populate(test_run.plan.product_id)
+
+        return {
+            'test_run': test_run,
+            'test_plan': test_run.plan,
+            'form': form,
+        }
+
+    @classmethod
+    def _get_test_run_or_fail(cls, run_id):
+        try:
+            test_run = TestRun.objects.select_related().get(run_id=run_id)
+        except ObjectDoesNotExist:
+            raise Http404
+
+        return test_run
+
+    @classmethod
+    def _get_base_form(cls, test_run):
```

Comment:
What is the difference between `_get_base_form()` and `get_form()` ? Why do we need the 2 when they look very similar ?

Reply:
Just want to move duplicate logic in one place, therefore I created _get_base_form. The get_form populate the form with data from the request object.